beacon-chain/sync/data_column_sidecars.go

```diff
@@ -39,249 +39,394 @@ type DataColumnSidecarsParams struct {
-	indicesMap map[uint64]bool,
-) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, error) {
-	if len(roBlocks) == 0 || len(indicesMap) == 0 {
-		return nil, nil
+	requestedIndices map[uint64]bool,
+) (map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn, map[[fieldparams.RootLength]byte]map[uint64]bool, error) {
+	if len(roBlocks) == 0 || len(requestedIndices) == 0 {
+		return nil, nil, nil
 	}
 
-	indices := sortedSliceFromMap(indicesMap)
-	slotsWithCommitments := make(map[primitives.Slot]bool)
-	missingIndicesByRoot := make(map[[fieldparams.RootLength]byte]map[uint64]bool)
-	indicesByRootStored := make(map[[fieldparams.RootLength]byte]map[uint64]bool)
-	result := make(map[[fieldparams.RootLength]byte][]blocks.VerifiedRODataColumn)
+	// Extract slots and roots corresponding to blocks with commitments.
+	slotsWithCommitments, storedIndicesByRoot, incompleteRoots, err := commitmentsInfo(params.Storage, roBlocks)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "commitments info")
+	}
 
-	for _, roBlock := range roBlocks {
-		// Filter out blocks without commitments.
-		block := roBlock.Block()
-		commitments, err := block.Body().BlobKzgCommitments()
-		if err != nil {
-			return nil, errors.Wrapf(err, "get blob kzg commitments for block root %#x", roBlock.Root())
-		}
-		if len(commitments) == 0 {
-			continue
-		}
+	initialMissingRootCount := len(incompleteRoots)
 
-		slotsWithCommitments[block.Slot()] = true
-		root := roBlock.Root()
+	// Request sidecars from storage (by reconstructing them from other available sidecars if needed).
+	result, err := requestSidecarsFromStorage(params.Storage, incompleteRoots, storedIndicesByRoot, requestedIndices)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "request sidecars from storage")
+	}
 
-		// Step 1: Get the requested sidecars for this root if available in storage
-		requestedColumns, err := tryGetStoredColumns(params.Storage, root, indices)
-		if err != nil {
-			return nil, errors.Wrapf(err, "try get direct columns for root %#x", root)
-		}
-		if requestedColumns != nil {
-			result[root] = requestedColumns
-			continue
-		}
+	log := log.WithField("initialMissingRootCount", initialMissingRootCount)
 
-		// Step 2: If step 1 failed, reconstruct the requested sidecars from what is available in storage
-		requestedColumns, err = tryGetReconstructedColumns(params.Storage, root, indices)
-		if err != nil {
-			return nil, errors.Wrapf(err, "try get reconstructed columns for root %#x", root)
-		}
-		if requestedColumns != nil {
-			result[root] = requestedColumns
-			continue
-		}
+	if len(incompleteRoots) == 0 {
+		log.WithField("finalMissingRootCount", len(incompleteRoots)).Debug("Fetched data column sidecars from storage")
+		return result, nil, nil
+	}
+
+	// Request direct sidecars from peers.
+	directSidecarsByRoot, err := requestDirectSidecarsFromPeers(params, roBlocks, slotsWithCommitments, storedIndicesByRoot, requestedIndices, incompleteRoots)
```

Comment:
also wondering we do 
1. request from storage
2. request from peers
could we possibly have gotten more into storage between these 2 checks?

Reply:
That's theoretically possible. However, it would mean Prysm is quite inefficient because, in the first place, it calls `FetchDataColumnSidecars` concurrently for the same blocks.

IMO it's an edge case not worth to deal with.